### PR TITLE
filter out secrets of command for rollout

### PR DESCRIPTION
**Pull request by xilabao**
this pr is used to hide user's secret information from "kubectl rollout histroy ..."

The current is someone create a pod specified the username and password, others can see the secrets.
for example:

```
ubuntu@kube-master:~$ kubectl create -f nginx.yaml --record -s https://127.0.0.1:6443 --insecure-skip-tls-verify=true --username=test --password=123456
deployment "my-nginx" created

ubuntu@kube-master:~$ kubectl rollout history deployment/my-nginx
deployments "my-nginx"
REVISION    CHANGE-CAUSE
1       kubectl create -f nginx.yaml --record -s https://127.0.0.1:6443 --insecure-skip-tls-verify=true --username=test--password=123456
```

username&password will be exposed
so we can hide token、username and password, such as:

use username/password

```

ubuntu@kube-master:~$ kubectl create -f nginx.yaml --record -s https://127.0.0.1:6443 --insecure-skip-tls-verify=true --username=test --password=123456
deployment "my-nginx" created

ubuntu@kube-master:~$ kubectl rollout history deployment/my-nginx
deployments "my-nginx"
REVISION    CHANGE-CAUSE
1       kubectl create -f nginx.yaml --record -s https://127.0.0.1:6443 --insecure-skip-tls-verify=true --username=****** --password=******

```

use token

```
ubuntu@kube-master:~$ kubectl edit deployment/my-nginx  -s https://127.0.0.1:6443 --insecure-skip-tls-verify=true --token=123456
deployment "my-nginx" edited

ubuntu@kube-master:~$ kubectl rollout history deployment/my-nginx
deployments "my-nginx"
REVISION    CHANGE-CAUSE
1       kubectl create -f nginx.yaml --record -s https://127.0.0.1:6443 --insecure-skip-tls-verify=true --username=****** --password=******
2       kubectl edit deployment/my-nginx -s https://127.0.0.1:6443 --insecure-skip-tls-verify=true --token=******
```

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35030)

### CODE REVIEW

**Comment by xilabao** [comment]
@k8s-bot ok to test

**Comment by kargakis** [comment]
@kubernetes/kubectl

**Comment by smarterclayton** [comment]
This is unfortunately not all flags that could be secrets.  The best option would be to add support in `spf13/pflag` to tag a flag as a secret, and then use that bit to determine the list.

Also, `Command()` could be used in contexts that need exact parameters (for subshell execution), so we would need to add a new method or extend the signature here to allow exact flags to be retrieved.

**Comment by xilabao** [comment]
@smarterclayton you are right. I can volunteer to finish the feature if you haven't already started on that.

File: pkg/kubectl/cmd/util/factory.go [hunk 1]
```diff
@@ -940,16 +940,44 @@ func GetFirstPod(client coreclient.PodsGetter, namespace string, selector labels
...
-func (f *factory) Command() string {
+func (f *factory) Command(enableSecret bool) string {
 	if len(os.Args) == 0 {
 		return ""
 	}
 	base := filepath.Base(os.Args[0])
 	args := append([]string{base}, os.Args[1:]...)
+
+	for i := 0; i < len(args); i++ {
+		if strings.HasPrefix(args[i], "--") {
```
Comment (eparis) [inline pkg/kubectl/cmd/util/factory.go]: This does not handle 'short' flags. Which stinks. Should we find a way to library-ify the command line parsing logic in pflag? `spf13/pflag/flags.go::parseArgs()` is quite complex to get right...

File: pkg/kubectl/cmd/util/factory.go [hunk 2]
```diff
@@ -940,16 +940,44 @@ func GetFirstPod(client coreclient.PodsGetter, namespace string, selector labels
...
+					return ""
+				}
+				args[i] = "--" + name + "=" + value
+			}
+		}
+	}
 	return strings.Join(args, " ")
 }
 
+func GetFlagValue(flags *pflag.FlagSet, name string, enableSecret bool) (string, error) {
```
Comment (eparis) [inline pkg/kubectl/cmd/util/factory.go]: I do not believe this function serves any value....

**Comment by eparis** [comment]
If you want to solve this, generically, in cobra/pflag we can, but we don't "need" to. Solving it generically in cobra/pflag would mean adding a function like "GetCmdLine()" or something to cobra. That would run all of the flags in the flagset, ignoring those which were unset, and printing those which were set. We would stop doing `os.Args` parsing at all in kubernetes (maybe a good thing). Given that the flags are in a map we are going to lose the ordering (and maybe predictability)...  which stinks...

But it would mean that cobra/pflag would know you are trying to get the command line which would mean that things like 'secret' would make sense to those tools. They would do some ****** type transformation before printing. Normally cobra/pflag would have no idea when the calling code wants the actual value or ******. Which is why i've been hesitant of your work in pflag. I've never understood how pflag was supposed to know what to do.

File: pkg/kubectl/cmd/util/factory.go [hunk 3]
```diff
@@ -940,16 +940,44 @@ func GetFirstPod(client coreclient.PodsGetter, namespace string, selector labels
...
+			name := (args[i])[2:]
+			if len(name) == 0 || name[0] == '-' || name[0] == '=' {
+				return ""
+			}
+
+			split := strings.SplitN(name, "=", 2)
+			name = split[0]
+
+			flag := f.flags.Lookup(name)
+			if flag != nil && flag.Secret {
```
Comment (eparis) [inline pkg/kubectl/cmd/util/factory.go]: I do not think that we need `flag.Secret` / changes to pflag. Instead set an annotation on the flag back when it was created and check for that annotation here.

File: pkg/kubectl/cmd/util/factory.go [hunk 4]
```diff
@@ -940,16 +940,44 @@ func GetFirstPod(client coreclient.PodsGetter, namespace string, selector labels
...
+				value, err := GetFlagValue(f.flags, name, enableSecret)
```
Comment (eparis) [inline pkg/kubectl/cmd/util/factory.go]: Why do we need `GetFlagValue()` ?
```
If enableSecret {
    args[i] = "--" + name + "=***********"
} else {
   // do nothing, args already fine
}
````

File: pkg/kubectl/cmd/util/factory.go [hunk 5]
```diff
@@ -940,16 +940,44 @@ func GetFirstPod(client coreclient.PodsGetter, namespace string, selector labels
 	return pod, 1, nil
 }
 
-// TODO: We need to filter out stuff like secrets.
-func (f *factory) Command() string {
+func (f *factory) Command(enableSecret bool) string {
```
Comment (eparis) [inline pkg/kubectl/cmd/util/factory.go]: maybe `printSecret`, `displaySecret`, or `showSecret` instead? Is there ever a case where we'd want to display the secret? Do we need to have that option?

**Review by eparis — CHANGES_REQUESTED** [review]
You can take 2 paths.  1 replace all of this with a new function in Cobra/pflag.

Follow these comments to simplify things here. I don't love that "short flags" are ignored, but its a step in a better direction....

File: pkg/kubectl/cmd/util/factory.go (hunk 1 above)
Comment (xilabao) [inline pkg/kubectl/cmd/util/factory.go]: @eparis I need to get flag.NoOptDefVal to make a distinction between "--flag" and "--flag arg". I can use flags.Lookup() to handle 'long' flags. but I can't find a way to handle 'short' flag. Do I ignore something?

**Comment by xilabao** [comment]
@eparis After the work (https://github.com/spf13/pflag/pull/107) has been done, I update the patch to use new function. PTAL.

**Comment by xilabao** [comment]
The new result will be
```
# ./cluster/kubectl.sh create --record=true -f ~/nginx.yaml --token=test/system:masters
deployment "my-nginx" created
# ./cluster/kubectl.sh  rollout history deployment/my-nginx                            
deployments "my-nginx"
REVISION	CHANGE-CAUSE
1		kubectl --record true --filename /home/nfs/nginx.yaml --token XXX
```

**Pull request closed without merging**